File: C++/DataStructures/LinkedList/SinglyLinkedList.cpp
// LinkedList.cpp : This file contains the 'main' function. Program execution begins and ends there.
//

#include <iostream>
using namespace std;
class Node {
public:
	int data;
	Node* next;
};
void push(Node** head, int newData) {
	Node* newNode = new Node();
	newNode->data = newData;
	newNode->next = *head;
	*head = newNode;
}

void insertAfter(Node* prevNode, int newData) {
	if (prevNode == NULL) {
		cout << "The give previous nde can't be NULL";
		return;
	}

	Node* newNode = new Node();
	newNode->data = newData;
	newNode->next = prevNode->next;
	prevNode->next = newNode;
}

void append(Node** head, int newData) {
	Node* newNode = new Node();
	Node* last = *head;
	newNode->data = newData;
	newNode->next = NULL;

	if (*head == NULL) {
		*head = newNode;
		return;
	}

	while (last->next != NULL) {
		last = last->next;
	}

	last->next = newNode;
	return;
}

void deleteNode(Node** head, int key) {
	Node* temp = *head;
	Node* prev = NULL;

	if (temp != NULL && temp->data == key) {
		*head = temp->next;

		delete temp;
		return;
	}

	else {
		while (temp != NULL && temp->data != key) {
			prev = temp;
			temp = temp->next;
		}

		if (temp == NULL)return;

		prev->next = temp->next;

		delete temp;
	}
}

int search(Node** head, int key) {
	Node* temp = *head;
	int index = 0;
	while (temp->data != key && temp->next != NULL) {
		index++;
		temp = temp->next;
	}

	if (temp->data != key)
		return -1;
	return index;
}

void display(Node* node) {

	while (node != NULL) {
		cout << node->data << " ";
		node = node->next;
	}

}


int main()
{
	Node* head = NULL;

	append(&head, 5);

	push(&head, 6);

	push(&head, 1);

	append(&head, 4);

	insertAfter(head->next, 8);

	cout << "Linked list is: ";
	display(head);

	deleteNode(&head, 1);
	cout << "\nLinked List after Deletion of 1: ";
	display(head);

	cout << "\nIs 22 in Linked list? ";
	int index = search(&head, 22);
		index !=-1 ? cout << "Yes, at the index "<<index : cout << "No";
	cout << "\nIs 6 in Linked list? ";
	index = search(&head, 6);
	index !=-1 ? cout << "Yes, at the index "<<index : cout << "No";

	return 0;
}

File: Python/BasicDataStructures/SinglyLinkedList.py
class Node():
    def __init__(self, data):
        self.data = data
        self.next = None
class LinkedList:
    def __init__(self):
        self.head = None
    def printList(self):
        temp = self.head
        while temp:
            print(temp.data,end=" ")
            temp = temp.next
    def insertFirst(self, x):
        newNode = Node(x)
        newNode.next = self.head
        self.head = newNode
    def insertAfter(self, w,x):
        if w is None:
            print("The given previous node must be in linkedList.")
            return
        newNode=Node(x)
        newNode.next = w.next
        w.next = newNode
    def insertLast(self, x):
        newNode=Node(x)
        if self.head is None:
            self.head = newNode
            return
        last = self.head
        while(last.next):
            last = last.next
        last.next = newNode
    def deleteFirst(self):
        if not self.head:
            return None
        temp = self.head
        self.head = self.head.next
        return temp
    def deleteLast(self):
        if self.head is None:
            return None
        if self.head.next is None:
            temp = self.head
            self.head = None
            return temp
        secondLast = self.head
        while secondLast.next.next:
            secondLast = secondLast.next
        temp = secondLast.next
        secondLast.next = None
        return temp
        
    def delete(self, x):
        y = self.head
        if y is not None:
            if y.data == x:
                self.head = y.next
                y = None
                return
        while(y is not None):
            if y.data == x: break
            w = y
            y = y.next
        if y is None: return
        w.next = y.next
        y = None
    def search(self, x):
        y = self.head
        index = 0
        while (y.data != x and y.next is not None):
            index+=1
            y = y.next

        if (y.data != x):
            return False
        return index

        
        
list = LinkedList()
list.insertLast(6)
list.insertFirst(7)
list.insertFirst(1)
list.insertLast(4)
list.insertAfter(list.head.next, 8)
list.delete(8)
list.deleteFirst()
list.deleteLast()
print(list.search(4))
list.printList()
